spec/syntax.md

```diff
@@ -906,16 +906,28 @@ backslash       = %x5C ; U+005C REVERSE SOLIDUS "\"
 ### Whitespace
 
 **_<dfn>Whitespace</dfn>_** is defined as one or more of
-U+0009 CHARACTER TABULATION (tab), 
-U+000A LINE FEED (new line),
-U+000D CARRIAGE RETURN, 
-U+3000 IDEOGRAPHIC SPACE, 
-or U+0020 SPACE.
+`U+0009 CHARACTER TABULATION` (tab), 
+`U+000A LINE FEED` (new line),
+`U+000D CARRIAGE RETURN`, 
+`U+3000 IDEOGRAPHIC SPACE`, 
+or `U+0020 SPACE`,
+optionally prepended with `U+200E LEFT-TO-RIGHT MARK`.
 
 Inside _patterns_ and _quoted literals_,
 whitespace is part of the content and is recorded and stored verbatim.
 Whitespace is not significant outside translatable text, except where required by the syntax.
 
+The character `U+200E LEFT-TO-RIGHT MARK` (LRM) MAY be prepended to _whitespace_ outside
+_patterns_ and _quoted literals_ to assist with presentation to users.
+Tools SHOULD generate these LRM characters following _identifiers_, _unquoted literals_, or 
+_option_ values that use right-to-left characters so that the _message_ displays
+intelligibly in a left-to-right context.
+
+This definition of _whitespace_ implements 
+[UTR#31 Rule 3a-2](https://www.unicode.org/reports/tr31/#R3a-2).
+It is a profile of R3a-1 in that specification because only the
+whitespace characters listed are permitted as whitespace.
```

Comment:
UAX31-R2a-2 says you need to
> define that profile with a precise specification of the characters that are added to or removed from the set of code points defined by the Pattern_White_Space property, and of any changes to the criteria under which a character or sequence of characters is interpreted as an end of line, as ignorable format controls, or as horizontal space.

The point is that the reader of such a conformance statement sees the difference from the default, which are the things that may need special attention when interoperating with an implementation based on the defaults.

So, if I am reading this right:
- Form feed, next line, line separator, paragraph separator, and right-to-left mark are removed from the set of white space characters;
- ideographic space is added to the set of white space characters;
- the sole ignorable format control (LRM) is allowed only in contexts UAX31-I1 and UAX31-I3 (not UAX31-I2), where UAX31-I1 is further restricted to the beginning of a sequence of horizontal spaces, and UAX31-I3 is further restricted to only the end and not the beginning of a line.

Reply:
> further restricted to only the end and not the beginning of a line

Also it is disallowed on a blank line, I think?

Reply:
> Note that quoted literals or pattern text can contain bidi controls that might cause Trojan source effects unless/until we address placeholder isolation.

Mostly that is something to be dealt with at a higher level (in editors and tooling); the main thing languages need to do is to treat the right things in the right way so that standardized tooling can deal with the issue, see UTS55.

> Note that UAX31-I2 is not permitted because that would break our sigil-identifier syntax.

I don’t really understand what you mean here; UAX31-I2 does not say you should allow `$⟨LRM⟩identifier`, it says that you should allow it _if_ you allow `$    identifier`.

The reason why you do not have UAX31-I2 is that you do not allow an LRM wherever you have `[s]`, so for instance in 

https://github.com/unicode-org/message-format-wg/blob/ae712d70d60e9b5d06bc72d5c7788811aa728f8a/spec/message.abnf#L35-L36

you do not allow `{⟨LRM⟩#whatever⟨LRM⟩}`, only `{⟨LRM⟩ #whatever⟨LRM⟩ }`, right?

Reply:
> you do not allow {⟨LRM⟩#whatever⟨LRM⟩}, only {⟨LRM⟩ #whatever⟨LRM⟩ }, right?

Note that this probably throws a wrench into the UTS55 conversion to plain text algorithm.

Reply:
Yes. See my response to @gibson042's comment.